operator/validator/metadata/syncer.go

```diff
@@ -0,0 +1,288 @@
+package metadata
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/attestantio/go-eth2-client/spec/phase0"
+	spectypes "github.com/ssvlabs/ssv-spec/types"
+	"go.uber.org/zap"
+
+	"github.com/ssvlabs/ssv/logging/fields"
+	"github.com/ssvlabs/ssv/protocol/v2/blockchain/beacon"
+	ssvtypes "github.com/ssvlabs/ssv/protocol/v2/types"
+	registrystorage "github.com/ssvlabs/ssv/registry/storage"
+	"github.com/ssvlabs/ssv/storage/basedb"
+)
+
+//go:generate mockgen -package=metadata -destination=./mocks.go -source=./syncer.go
+
+const (
+	defaultSyncInterval      = 12 * time.Minute
+	defaultStreamInterval    = 2 * time.Second
+	defaultUpdateSendTimeout = 30 * time.Second
+	batchSize                = 512
+)
+
+type ValidatorSyncer struct {
+	logger            *zap.Logger
+	shareStorage      shareStorage
+	beaconNetwork     beacon.BeaconNetwork
+	fetcher           fetcher
+	syncInterval      time.Duration
+	streamInterval    time.Duration
+	updateSendTimeout time.Duration
+}
+
+type shareStorage interface {
+	List(txn basedb.Reader, filters ...registrystorage.SharesFilter) []*ssvtypes.SSVShare
+	Range(txn basedb.Reader, fn func(*ssvtypes.SSVShare) bool)
+	UpdateValidatorsMetadata(map[spectypes.ValidatorPK]*beacon.ValidatorMetadata) error
+}
+
+type fetcher interface {
+	Fetch(ctx context.Context, pubKeys []spectypes.ValidatorPK) (Validators, error)
+}
+
+func NewValidatorSyncer(
+	logger *zap.Logger,
+	shareStorage shareStorage,
+	beaconNetwork beacon.BeaconNetwork,
+	beaconNode beacon.BeaconNode,
+	opts ...Option,
+) *ValidatorSyncer {
+	u := &ValidatorSyncer{
+		logger:            logger,
+		shareStorage:      shareStorage,
+		beaconNetwork:     beaconNetwork,
+		fetcher:           NewFetcher(logger, beaconNode),
+		syncInterval:      defaultSyncInterval,
+		streamInterval:    defaultStreamInterval,
+		updateSendTimeout: defaultUpdateSendTimeout,
+	}
+
+	for _, opt := range opts {
+		opt(u)
+	}
+
+	return u
+}
+
+type Option func(*ValidatorSyncer)
+
+func WithSyncInterval(interval time.Duration) Option {
+	return func(u *ValidatorSyncer) {
+		u.syncInterval = interval
+	}
+}
+
+func (u *ValidatorSyncer) SyncOnStartup(ctx context.Context) (map[spectypes.ValidatorPK]*beacon.ValidatorMetadata, error) {
+	// Load non-liquidated shares.
+	shares := u.shareStorage.List(nil, registrystorage.ByNotLiquidated())
+	if len(shares) == 0 {
+		u.logger.Info("could not find non-liquidated validator shares on initial metadata retrieval")
+		return nil, nil
+	}
+
+	needToSync := false
+	allPubKeys := make([]spectypes.ValidatorPK, 0, len(shares))
+	for _, share := range shares {
+		allPubKeys = append(allPubKeys, share.ValidatorPubKey)
+		if !share.HasBeaconMetadata() {
+			needToSync = true
```

Comment:
you can break the loop here

Reply:
@y0sher This is two loops combined in one: to get all pubkeys and to check if need to sync, so I cannot break it because I need all pubkeys. Do you think having two loops and breaking the second one would be more readable?

Reply:
nah now I understand its ok